src/index.ts

```diff
@@ -83,77 +75,114 @@ export function latexRequest<T>(
-          errorPanel.text = err.xhr.response;
-        });
-      });
 
-      // Run an initial latexRequest so that the appropriate files exist,
-      // then open them.
-      latexRequest(widget.context.path, serverSettings).then(() => {
-        // Open the pdf and get a handle on its document context.
-        const pdfWidget = manager.openOrReveal(pdfFilePath);
-        pdfContext = manager.contextForWidget(pdfWidget);
+  // Given an fileEditor widget that hosts
+  // a .tex document, open a LaTeX preview for it.
+  const openPreview = (widget: DocumentRegistry.IReadyWidget) => {
+    // If we can't find the document context, bail.
+    let texContext = manager.contextForWidget(widget);
+    if (!texContext) {
+      return;
+    }
+    // If there is already an active preview for this context, bail.
+    if (Private.previews.has(texContext.path)) {
+      return;
+    }
+
+    // build pdfFilePath so that we know what to watch for
+    const dirName = PathExt.dirname(texContext.path);
+    const baseName = PathExt.basename(texContext.path, '.tex');
+    const pdfFilePath = PathExt.join(dirName, baseName + '.pdf');
+
+    let pdfContext: DocumentRegistry.IContext<DocumentRegistry.IModel>;
+    let errorPanel: ErrorPanel | null = null;
+
+    // Hook up an event listener for when the '.tex' file is saved.
+    texContext.fileChanged.connect((sender, args) => {
+      latexRequest(texContext.path, serverSettings).then(() => {
+        // Read the pdf file contents from disk.
+        if (pdfContext) {
+          pdfContext.revert();
+        } else {
+          const pdfWidget = manager.openOrReveal(pdfFilePath);
+          pdfContext = manager.contextForWidget(pdfWidget);
+        }
+        if (errorPanel) {
+          errorPanel.close();
+        }
       }).catch((err) => {
         // If there was an error, read the log
         // file from disk and show it.
-        errorPanel = new ErrorPanel();
+        if (!errorPanel) {
+          errorPanel = Private.createErrorPanel();
+          // On disposal, set the reference to null
+          errorPanel.disposed.connect(() => {
+            errorPanel = null;
+          });
+          //Add the error panel to the main area.
+          app.shell.addToMainArea(errorPanel, { ref: widget.id });
+        }
         errorPanel.text = err.xhr.response;
-        errorPanel.id = `latex-error-${++Private.id}`;
-        errorPanel.title.label = 'LaTeX Error';
-        errorPanel.title.closable = true;
-        app.shell.addToMainArea(errorPanel, { ref: widget.id });
       });
+    });
+
+    // Run an initial latexRequest so that the appropriate files exist,
+    // then open them.
+    latexRequest(texContext.path, serverSettings).then(() => {
+      // Open the pdf and get a handle on its document context.
+      const pdfWidget = manager.openOrReveal(pdfFilePath);
+      pdfContext = manager.contextForWidget(pdfWidget);
+    }).catch((err) => {
+      // If there was an error, read the log
```

Comment:
This is no longer accurate… but the fact that this error appears elsewhere suggests that there might be some further deduplication that we can do in a later PR.

Reply:
Right you are! We may be able to combine the initial link with the on-saved event if we restore the `context.save()` at the end.